src/Invitation.ts

```diff
@@ -0,0 +1,201 @@
+      invitation.SealedInvitationHeaderV1.decode(bytes)
+    )
+  }
+}
+
+export class SealedInvitationV1 implements invitation.SealedInvitationV1 {
+  headerBytes: Uint8Array
+  ciphertext: Ciphertext
+  private _header?: SealedInvitationHeaderV1
+  private _invitation?: InvitationV1
+
+  constructor({ headerBytes, ciphertext }: invitation.SealedInvitationV1) {
+    if (!headerBytes || !headerBytes.length) {
+      throw new Error('Missing header bytes')
+    }
+    if (!ciphertext) {
+      throw new Error('Missing ciphertext')
+    }
+    this.headerBytes = headerBytes
+    this.ciphertext = new Ciphertext(ciphertext)
+  }
+
+  /**
+   * Accessor method for the full header object
+   */
+  get header(): SealedInvitationHeaderV1 {
+    // Use cached value if already exists
+    if (this._header) {
+      return this._header
+    }
+    this._header = SealedInvitationHeaderV1.fromBytes(this.headerBytes)
+    return this._header
+  }
+
+  /**
+   * getInvitation decrypts and returns the InvitationV1 stored in the ciphertext of the Sealed Invitation
+   */
+  async getInvitation(viewer: PrivateKeyBundleV2): Promise<InvitationV1> {
+    // Use cached value if already exists
+    if (this._invitation) {
+      return this._invitation
+    }
+    // The constructors for child classes will validate that this is complete
+    const header = this.header
+    let secret: Uint8Array
+    if (viewer.identityKey.matches(this.header.sender.identityKey)) {
+      secret = await viewer.sharedSecret(
+        header.recipient,
+        header.sender.preKey,
+        false
+      )
+    } else {
+      secret = await viewer.sharedSecret(
+        header.sender,
+        header.recipient.preKey,
+        true
+      )
+    }
+
+    const decryptedBytes = await decrypt(
+      this.ciphertext,
+      secret,
+      this.headerBytes
+    )
+    this._invitation = InvitationV1.fromBytes(decryptedBytes)
+    return this._invitation
+  }
+
+  toBytes(): Uint8Array {
+    return invitation.SealedInvitationV1.encode(this).finish()
+  }
+
+  static fromBytes(bytes: Uint8Array): SealedInvitationV1 {
+    return new SealedInvitationV1(invitation.SealedInvitationV1.decode(bytes))
+  }
+}
+
+/**
+ * Wrapper class for SealedInvitationV1 and any future iterations of SealedInvitation
+ */
+export class SealedInvitation implements invitation.SealedInvitation {
+  v1: SealedInvitationV1
+
+  constructor({ v1 }: invitation.SealedInvitation) {
+    if (!v1) {
+      throw new Error('Missing v1 invitation')
+    }
+    this.v1 = new SealedInvitationV1(v1)
+  }
+
+  toBytes(): Uint8Array {
+    return invitation.SealedInvitation.encode(this).finish()
+  }
+
+  static fromBytes(bytes: Uint8Array): SealedInvitation {
+    return new SealedInvitation(invitation.SealedInvitation.decode(bytes))
+  }
+
+  /**
+   * Create a SealedInvitation with a SealedInvitationV1 payload
+   * Will encrypt all contents and validate inputs
+   */
+  static async createV1({
```

Comment:
This is just a nitpick, but would this make more sense as `SealedInvitationV1.create()` ?

Reply:
My thinking is that I don't want all consumers have to have to do `new SealedInvitation({ v1: await SealedInvitationV1.create(...) })`. Feels like no one wants the V1 invitation, and everyone will want the actual proto that is going over the wire.